paccmann_chemistry/utils/utils.py

```diff
@@ -67,6 +68,71 @@ def sequential_data_preparation(
     return input_seq.to(device), decoder_seq.to(device), target_seq.to(device)
 
 
+def packed_sequential_data_preparation(
+    input_batch,
+    input_keep=1,
+    start_index=2,
+    end_index=3,
+    dropout_index=1,
+):
+    """
+    Sequential Training Data Builder.
+
+    Args:
+        input_batch (torch.Tensor): Batch of padded sequences, output of
+            nn.utils.rnn.pad_sequence(batch) of size
+            `[sequence length, batch_size, 1]`.
+        input_keep (float): The probability not to drop input sequence tokens
+            according to a Bernoulli distribution with p = input_keep.
+            Defaults to 1.
+        start_index (int): The index of the sequence start token.
+        end_index (int): The index of the sequence end token.
+        dropout_index (int): The index of the dropout token. Defaults to 1.
+
+    Returns:
+    (torch.Tensor, torch.Tensor, torch.Tensor): encoder_seq, decoder_seq,
+        target_seq
+
+        encoder_seq is a batch of padded input sequences starting with the
+            start_index, of size `[sequence length +1, batch_size, 1]`.
+        decoder_seq is like encoder_seq but word dropout is applied
+            (so if input_keep==1, then decoder_seq = encoder_seq).
+        target_seq (torch.Tensor): Batch of padded target sequences ending
+            in the end_index, of size `[sequence length +1, batch_size, 1]`.
+    """
+
+    def _process_sample(sample):
+        if len(sample.shape) != 1:
+            raise ValueError
+        input = torch.LongTensor(sample.cpu().numpy())
+        decoder = input.clone()
+
+        # apply token dropout if keep != 1
+        if input_keep != 1:
+            # mask for token dropout
+            mask = Bernoulli(input_keep).sample(input.shape)
+            mask = torch.LongTensor(mask.numpy())
+            dropout_loc = np.where(mask == 0)[0]
+            decoder[dropout_loc] = dropout_index
+
+        # TODO: Don't we have the start index at the begining already?
+        # input_seq = torch.cat((torch.tensor([start_index]), input))
+        # decoder_seq = torch.cat((torch.tensor([start_index]), decoder))
```

Comment:
I left this TODO here when I merged, and no I am not sure if what I was doing here is correct. @jannisborn I did remove this part here because the sequences I was passing in already had the start and end index. We should double check that and if the start/end token are already there we can remove them from this fn (or if they are not un-comment this part). In any case, there may be an inconsistency with the `padded` data prep.

Reply:
you're right. This should be controlled only in the `SMILESDataset` and the `SMILESLanguage`. I removed the addition of the start_index from both data preparation methods and also refactored/speeded it up a bit.